File: iot/amqp/terminus/AMQPTarget.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.avps.TerminusDurability as TerminusDurability
import iot.amqp.avps.TerminusExpiryPolicy as TerminusExpiryPolicy
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.tlv.impl.TLVList as TLVList
import iot.amqp.tlv.impl.TLVFixed as TLVFixed

class amqpTarget():
    def __init__(self,address,durable,expiryPeriod,timeout,dynamic,dynamicNodeProperties,capabilities):
        self.address = address
        self.durable = durable
        self.expiryPeriod = expiryPeriod
        self.timeout = timeout
        self.dynamic = dynamic
        self.dynamicNodeProperties = dynamicNodeProperties
        self.capabilities = capabilities
        self.amqpType = AMQPType.amqpType()

    def toArgumentsList(self):
        list = TLVList.tlvList(None,None)
        wrapper = AMQPWrapper.amqpWrapper()
        if self.address is not None:
            list.addElement(0, wrapper.wrap(self.address))

        if self.durable is not None:
            list.addElement(1, wrapper.wrap(self.durable))
        if self.expiryPeriod is not None:
            list.addElement(2, wrapper.wrap(self.expiryPeriod))
        if self.timeout is not None:
            result = wrapper.wrap(self.timeout)
            list.addElement(3, result)
        if self.dynamic is not None:
            list.addElement(4, wrapper.wrap(self.dynamic))
        if self.dynamicNodeProperties is not None:
            if self.dynamic is not None:
                if self.dynamic:
                    list.addElement(5, wrapper.wrapMap(self.dynamicNodeProperties))
                else:
                    print("Target's dynamic-node-properties can't be specified when dynamic flag is false")
            else:
                print("STarget's dynamic-node-properties can't be specified when dynamic flag is not set")

        if self.capabilities is not None and len(self.capabilities) > 0:
            list.addElement(6, wrapper.wrapArray(self.capabilities))

        constructor = DescribedConstructor.describedConstructor(list.getCode(),TLVFixed.tlvFixed(self.amqpType.getValueByKey('SMALL_ULONG'), 0x29))
        list.setConstructor(constructor)
        return list

    def fromArgumentsList(self, list):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        if len(list.getList()) > 0 :
            element = list.getList()[0]
            if element is not None and not element.isNull():
                self.address = unwrapper.unwrapString(element)
        if len(list.getList()) > 1 :
            element = list.getList()[1]
            if element is not None and not element.isNull():
                self.durable = TerminusDurability.terminusDurability(unwrapper.unwrapUInt(element))
        if len(list.getList()) > 2 :
            element = list.getList()[2]
            if element is not None and not element.isNull():
                self.expiryPeriod = TerminusExpiryPolicy.terminusExpiryPolicy(unwrapper.unwrapSymbol(element))
        if len(list.getList()) > 3 :
            element = list.getList()[3]
            if element is not None and not element.isNull():
                self.timeout = unwrapper.unwrapUInt(element)
        if len(list.getList()) > 4 :
            element = list.getList()[4]
            if element is not None and not element.isNull():
                self.dynamic = unwrapper.unwrapBool(element)
        if len(list.getList()) > 5 :
            element = list.getList()[5]
            if element is not None and not element.isNull():
                if self.dynamic is not None:
                    self.dynamicNodeProperties = unwrapper.unwrapMap(element)
                else:
                    print("Received malformed Target: dynamic-node-properties can't be specified when dynamic flag is false")
            else:
                print("Received malformed Target: dynamic-node-properties can't be specified when dynamic flag is not set")
        if len(list.getList()) > 6:
            element = list.getList()[6]
            if element is not None and not element.isNull():
                self.capabilities = unwrapper.unwrapArray(element)

    def toString(self):
        return 'AMQPTarget [address=' + str(self.address) + ', durable=' + str(self.durable) + ', expiryPeriod=' + str(self.expiryPeriod) + ', timeout=' + str(self.timeout) + ', dynamic=' + str(self.dynamic) + ', dynamicNodeProperties=' + str(self.dynamicNodeProperties) + ', capabilities=' + str(self.capabilities) + ']'

    def getAddress(self):
        return self.address

    def setAddress(self, address):
        self.address = address

    def getDurable(self):
        return self.durable

    def setDurable(self, durable):
        self.durable = durable

    def getExpiryPeriod(self):
        return self.expiryPeriod

    def setExpiryPeriod(self, expiryPeriod):
        self.expiryPeriod = expiryPeriod

    def getTimeout(self):
        return self.timeout

    def setTimeout(self, timeout):
        self.timeout = timeout

    def getDynamic(self):
        return self.dynamic

    def setDynamic(self, dynamic):
        self.dynamic = dynamic

    def getDynamicNodeProperties(self):
        return self.dynamicNodeProperties

    def setDynamicNodeProperties(self, dynamicNodeProperties):
        self.dynamicNodeProperties = dynamicNodeProperties

    def getCapabilities(self):
        return self.capabilities

    def setCapabilities(self, capabilities):
        self.capabilities = capabilities

File: IoTbrokerClient/src/iot-protocols/amqp/classes/terminus/amqptarget.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqptarget.h"
#include "iot-protocols/amqp/classes/wrappers/amqpsymbol.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"
#include "iot-protocols/amqp/classes/tlv/api/amqptlvfactory.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvfixed.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpfactory.h"

AMQPTarget::AMQPTarget()
{
    this->expiryPeriod = NULL;
    this->dynamicNodeProperties = NULL;
    this->capabilities = NULL;
}

AMQPTLVList *AMQPTarget::getlist()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (!this->address.isEmpty()) {
        list->addElementWithIndex(0, AMQPWrapper::wrapString(this->address));
    }
    if (this->durable != NULL) {
        list->addElementWithIndex(1, AMQPWrapper::wrapUInt(this->durable->getValue()));
    }
    if (this->expiryPeriod != NULL) {
        AMQPSymbol *symbol = new AMQPSymbol(this->expiryPeriod->getName());
        list->addElementWithIndex(2, AMQPWrapper::wrapSymbol(symbol));
    }
    if (this->timeout != NULL) {
        list->addElementWithIndex(3, AMQPWrapper::wrapUInt(this->timeout));
    }
    if (this->dynamic != NULL) {
        list->addElementWithIndex(4, AMQPWrapper::wrapBool(this->dynamic));
    }
    if (this->dynamicNodeProperties.count() != 0) {
        if (this->dynamic != NULL) {
            if ((bool) this->dynamic) {
                list->addElementWithIndex(5, AMQPWrapper::wrapMap(this->dynamicNodeProperties));
            } else {
                printf("AMQPTarget::getList::dynamic");
            }
        } else {
            printf("AMQPTarget::getList::dynamicNodeProperties");
        }
    }

    if (this->capabilities.count() != 0) {
        list->addElementWithIndex(6, AMQPWrapper::wrapList(this->capabilities));
    }
    unsigned char * data = NULL;
    data = malloc(sizeof(char));
    data[0] = (char) 0x29;
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    list->setConstructor(constructor);

    return list;
}

void AMQPTarget::fill(AMQPTLVList *list)
{
    if (list->getList().size() > 0) {
        TLVAMQP *element = list->getList()[0];
        if (!element->isNull()) {
            this->address = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (list->getList().size() > 1) {
        TLVAMQP *element = list->getList()[1];
        if (!element->isNull()) {
            int number = AMQPUnwrapper::unwrapUInt(element);
            this->durable = new AMQPTerminusDurability(number);
        }
    }
    if (list->getList().size() > 2) {
        TLVAMQP *element = list->getList()[2];
        if (!element->isNull()) {
            AMQPSymbol *symbol = AMQPUnwrapper::unwrapSymbol(element);
            AMQPTerminusExpiryPolicy *expiryPolicy = new AMQPTerminusExpiryPolicy();
            expiryPolicy = new AMQPTerminusExpiryPolicy(expiryPolicy->EnumObject::getValue(symbol->getValue()));
            this->expiryPeriod = expiryPolicy;
        }
    }
    if (list->getList().size() > 3) {
        TLVAMQP *element = list->getList()[3];
        if (!element->isNull()) {
            JsonVariant variant;
            variant.set((unsigned int)AMQPUnwrapper::unwrapUInt(element));
            this->timeout = variant;
        }
    }
    if (list->getList().size() > 4) {
        TLVAMQP *element = list->getList()[4];
        if (!element->isNull()) {
            JsonVariant variant;
            variant.set(AMQPUnwrapper::unwrapBool(element));
            this->dynamic = variant;
        }
    }
    if (list->getList().size() > 5) {
        TLVAMQP *element = list->getList()[5];
        if (!element->isNull()) {
            if (this->dynamic != NULL) {
                if ((bool)this->dynamic) {
                    this->dynamicNodeProperties = AMQPUnwrapper::unwrapMap(element);
                } else {
                    printf("AMQPTarget::fill::dynamic");
                }
            } else {
                printf("AMQPTarget::fill::dynamicNodeProperties");
            }
        }
    }
    if (list->getList().size() > 6) {
        TLVAMQP *element = list->getList()[6];
        if (!element->isNull()) {
            this->capabilities = AMQPUnwrapper::unwrapList(element);
        }
    }
}

void AMQPTarget::addDynamicNodeProperties(String key, JsonVariant *value)
{
    JsonVariant *keyVariant;
    keyVariant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(key)));
    this->dynamicNodeProperties.insert(keyVariant, value);
}

void AMQPTarget::addCapabilities(list<String> array)
{
    foreach (String item, array) {
        JsonVariant *itemVariant;
        itemVariant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(item)));
        this->capabilities.push_back(itemVariant);
    }
}

String AMQPTarget::getAddress() const
{
    return address;
}

void AMQPTarget::setAddress(const String &value)
{
    address = value;
}

AMQPTerminusDurability *AMQPTarget::getDurable() const
{
    return durable;
}

void AMQPTarget::setDurable(AMQPTerminusDurability *value)
{
    durable = value;
}

AMQPTerminusExpiryPolicy *AMQPTarget::getExpiryPeriod() const
{
    return expiryPeriod;
}

void AMQPTarget::setExpiryPeriod(AMQPTerminusExpiryPolicy *value)
{
    expiryPeriod = value;
}

JsonVariant *AMQPTarget::getTimeout() const
{
    return timeout;
}

void AMQPTarget::setTimeout(JsonVariant *value)
{
    timeout = value;
}

JsonVariant *AMQPTarget::getDynamic() const
{
    return dynamic;
}

void AMQPTarget::setDynamic(JsonVariant *value)
{
    dynamic = value;
}

map<JsonVariant *, JsonVariant *> AMQPTarget::getDynamicNodeProperties() const
{
    return dynamicNodeProperties;
}

void AMQPTarget::setDynamicNodeProperties(const map<JsonVariant *, JsonVariant *> &value)
{
    dynamicNodeProperties = value;
}

list<JsonVariant *> AMQPTarget::getCapabilities() const
{
    return capabilities;
}

void AMQPTarget::setCapabilities(const list<JsonVariant *> &value)
{
    capabilities = value;
}
